feat: eth_sendRawTransactionConditional

hamdiallam opened the pull request:
Closes ethereum-optimism/ecopod#1024

Sequencer implementation of `eth_sendRawTransactionConditional` described in this [design doc](/ethereum-optimism/design-docs/blob/main/ecopod/sendRawTransactionConditional/proposal.md), refreshed from @tynes initial [draft](https://github.com/ethereum/go-ethereum/compare/master...tynes:go-ethereum:eip4337)

## Design Doc Divergence

Rather than exposing a separate port to expose the sequencer rpcs, we instead expand on the  `--rollup` configuration to specify whether or not this endpoint should be enabled. The default, `--rollup.sequencerenabletxconditional` is false, thus retaining the desired outcome that this endpoint must be explicitly enabled by the operator


## Extensions

Transactions with an attached conditional are bound to the block builder it was submitted to. This means that these transactions must be excluded from being gossiped to any configured peers via the p2p protocol.

## Metrics

Not all the metrics in the design doc are implemented here. In op-geth we primarily track mempool time and rejection. In the proxy, we'll track the request breakdown by authenticated caller and top-level success rate

The mempool latency is tracked by attaching the submission time to the conditional and recording the elapsed time when the transaction is mined or demoted from the txpool

hamdiallam commented on eth/protocols/eth/protocol.go:
```diff
@@ -111,9 +111,86 @@ func (p *NewBlockHashesPacket) Unpack() ([]common.Hash, []uint64) {
 	return hashes, numbers
 }
 
+// transactionEnvelope is an envelope over a transaction such that
```
See PR description open question about an alternative to this

protolambda commented on core/state/statedb.go:
```diff
@@ -368,6 +368,30 @@ func (s *StateDB) HasSelfDestructed(addr common.Address) bool {
 	return false
 }
 
+// CheckTransactionConditional validates the account preconditions against the statedb.
+func (s *StateDB) CheckTransactionConditional(cond *types.TransactionConditional) error {
+	for addr, acct := range cond.KnownAccounts {
+		if root, isRoot := acct.Root(); isRoot {
+			storageRoot := s.GetStorageRoot(addr)
```
Note that the state may change while this loop is running, since no lock is being held over the state-db, and it's not pinned to some snapshot either.

hamdiallam replied:
Added a comment to the docstring that the caller is responsible that no changes are made while this function is called. All uses are snapshots so we're good

protolambda commented on core/state/statedb.go:
```diff
@@ -368,6 +368,30 @@ func (s *StateDB) HasSelfDestructed(addr common.Address) bool {
 
+// CheckTransactionConditional validates the account preconditions against the statedb.
+func (s *StateDB) CheckTransactionConditional(cond *types.TransactionConditional) error {
+	for addr, acct := range cond.KnownAccounts {
```
It would be nice enforce some sanity limit on the number of conditional accounts / storage dependencies. E.g. if there's more than 100, don't start any work, and just return an error. To avoid accidental DoS.

I know the surface RPC enforces some limit, but conditional txs still come in through p2p and perhaps other routes. Good to have a check on the `cond.Cost()` where it most matters.

hamdiallam replied:
Sounds good, made the cost a public var in `types` that's also checked here

protolambda commented on cmd/utils/flags.go:
```diff
@@ -1866,6 +1872,11 @@ func SetEthConfig(ctx *cli.Context, stack *node.Node, cfg *ethconfig.Config) {
 	cfg.RollupHaltOnIncompatibleProtocolVersion = ctx.String(RollupHaltOnIncompatibleProtocolVersionFlag.Name)
 	cfg.ApplySuperchainUpgrades = ctx.Bool(RollupSuperchainUpgradesFlag.Name)
+
+	if ctx.IsSet(RollupSequencerEnableTxConditionalFlag.Name) {
+		cfg.RollupSequencerEnableTxConditional = ctx.Bool(RollupSequencerEnableTxConditionalFlag.Name)
+	}
```
It's fine to just rely on the boolean false default, no need to check if it's set first. See above `cfg.ApplySuperchainUpgrades` usage.

protolambda commented on core/txpool/legacypool/legacypool.go:
```diff
@@ -1739,9 +1742,26 @@ func (pool *LegacyPool) demoteUnexecutables() {
+
+			hash := tx.Hash()
+			pool.all.Remove(tx.Hash())
+			log.Trace("Removed stale transaction with invalidated conditional", "hash", hash)
```
```suggestion
			log.Debug("Removed stale transaction with invalidated conditional", "hash", hash)
```
I think this is worth bumping up, in case we're debugging missing txs.

hamdiallam replied:
I ended up removing this code since the miner will already check. No use in doing it twice on every new incoming block. However i bumped up the miner logs to Debug for this

protolambda commented on core/txpool/legacypool/legacypool.go:
```diff
@@ -1739,9 +1742,26 @@ func (pool *LegacyPool) demoteUnexecutables() {
+				txConditionalDemotedTimer.UpdateSince(conditional.SubmissionTime)
+			}
+
+			hash := tx.Hash()
+			pool.all.Remove(tx.Hash())
```
The `tx.Hash()` is cached inside of the tx object, but since it's used the very next line, it would be nice to just re-use the `hash` variable

protolambda commented on core/types/block.go:
```diff
@@ -165,6 +165,23 @@ func (h *Header) EmptyReceipts() bool {
+	if cond.BlockNumberMin != nil && cond.BlockNumberMin.Cmp(h.Number) > 0 {
+		return fmt.Errorf("failed block number minimum constraint")
+	}
+	if cond.BlockNumberMax != nil && cond.BlockNumberMax.Cmp(h.Number) < 0 {
+		return fmt.Errorf("failed block number maximmum constraint")
+	}
```
If BlockNumberMax == h.Number, then the comparison equals 0, and it's included. Is the upper bound meant to be inclusive? I'd expect an exclusive upper bound.

hamdiallam replied:
Per the [spec](https://notes.ethereum.org/@yoav/SkaX2lS9j) "blockNumberMax: [optional] maximum block number for inclusion"

I interpreted this as inclusive bounds (same working with timestamp bounds)

protolambda commented on core/types/transaction_conditional.go:
```diff
@@ -0,0 +1,256 @@
+	// Header state conditionals
+	BlockNumberMin *big.Int `json:"blockNumberMin,omitempty"`
+	BlockNumberMax *big.Int `json:"blockNumberMax,omitempty"`
+	TimestampMin   *uint64  `json:"timestampMin,omitempty"`
+	TimestampMax   *uint64  `json:"timestampMax,omitempty"`
```
Worth a comment if it's an inclusive or exclusive block number bound.

protolambda commented on core/types/block.go:
```diff
@@ -165,6 +165,23 @@ func (h *Header) EmptyReceipts() bool {
+	if cond.TimestampMin != nil && *cond.TimestampMin > h.Time {
+		return fmt.Errorf("failed timestamp minimum constraint")
+	}
+	if cond.TimestampMax != nil && *cond.TimestampMax < h.Time {
+		return fmt.Errorf("failed timestamp maximum constraint")
+	}
```
Same here, inclusive or exclusive?

protolambda commented on core/types/transaction.go:
```diff
@@ -387,6 +390,16 @@ func (tx *Transaction) RollupCostData() RollupCostData {
 	return out
 }
 
+// Conditional returns the conditional attached to the transaction
+func (tx *Transaction) Conditional() *TransactionConditional {
+	return tx.conditional
+}
+
+// SetConditional attaches a conditional to the transaction
+func (tx *Transaction) SetConditional(cond *TransactionConditional) {
+	tx.conditional = cond
+}
```
Generally it's not advised to mutate transaction attributes. But also I don't see a better option.

It would be nice to use a generic atomic pointer util though, to make the get/set thread-safe.

hamdiallam replied:
Ah yea makes sense to use an atomic pointer here, nice catch

protolambda commented on core/types/transaction_conditional.go:
```diff
@@ -0,0 +1,256 @@
+	return json.Marshal(ka.StorageSlots)
+}
+
+// EncodeRLP will serialize the KnownAccount into rlp bytes
+func (ka KnownAccount) EncodeRLP(w io.Writer) error {
```
I think a pointer-receiver here is better, since it embeds a map field, to not incur any accidental copy cost (assuming it's not optimized by Go, not sure)

protolambda commented on eth/protocols/eth/protocol.go:
```diff
@@ -111,9 +111,86 @@ func (p *NewBlockHashesPacket) Unpack() ([]common.Hash, []uint64) {
 
+// transactionEnvelope is an envelope over a transaction such that
+// out-of-protocol metadata can be gossiped to peers
+type transactionEnvelope struct {
+	Tx          *types.Transaction
+	Conditional *types.TransactionConditional
+}
```
Do we have specs for this? We need other EL clients to be able to sync transactions with op-geth still, if users start running op-reth/op-erigon/op-nethermind more in production sequencing settings.

protolambda commented on internal/sequencerapi/api.go:
```diff
@@ -0,0 +1,84 @@
+	}
+}
+
+func (s *sendRawTxCond) SendRawTransactionConditional(ctx context.Context, txBytes hexutil.Bytes, cond types.TransactionConditional) (common.Hash, error) {
+	sendRawTxConditionalRequestsCounter.Inc(1)
```
Does ERC-4337 or related work not describe error codes to use for certain validation conditions? By wrapping the returned eror with the geth RPC error, error-codes can be specified. At least wrapping it with standard input-errors, to distinguish from server errors, would be helpful to the end user.

tynes replied:
It does define specific error codes, the old implementation that I worked on defined some of them but not all of them

hamdiallam replied:
ah nice catch! Didn't realize there were error codes in the api. Went over the spec and added them. The json rpc error stuff isn't exported but straightforward to create my own error type here

protolambda commented on miner/worker.go:
```diff
@@ -799,6 +805,19 @@ func (w *worker) commitTransaction(env *environment, tx *types.Transaction) ([]*
 	if tx.Type() == types.BlobTxType {
 		return w.commitBlobTransaction(env, tx)
 	}
+
+	// If a conditional is set, check prior to applying
+	if conditional := tx.Conditional(); conditional != nil {
+		txConditionalMinedTimer.UpdateSince(conditional.SubmissionTime)
+
+		if err := env.header.CheckTransactionConditional(conditional); err != nil {
+			return nil, fmt.Errorf("failed tx conditional: %w", errTxConditionalInvalid)
+		}
+		if err := env.state.CheckTransactionConditional(conditional); err != nil {
+			return nil, fmt.Errorf("failed tx conditional: %w", errTxConditionalInvalid)
+		}
+	}
+
```
Can we add a block-building flag to rate-limit conditional-txs in the final synchronous block building work? Or some safety mechanism, in case the load is too much to sequence?

And ideally, a rate-limit expressed in condition-cost, not just number of txs. E.g. use the standard rate limit Go util, create a burst of X (set to maximum tx conditional cost), and throughput rate of Y (the configured rate), and then take from the bucket with the `cond.Cost()`. And don't block on rate limit, but just drop the tx if there's no capacity available, in favor of non-conditional txs. I think this would be a helpful anti-DoS measure.

hamdiallam replied:
Yeah good idea! Added a flag for the rate limit here. Set the default to be 5tps*max cost so 5000

protolambda commented:
Reviewed the implementation details and some DoS risks. Would also be helpful if a test plan is shared, and if someone from the node team can review also, because of the size of the PR.

And request: can you add a 4337 category to the `fork.yaml` file, and describe the changes? That way the op-geth.optimism.io website accurately represents the geth dfiff, and documents the 4337 work nicely. You can run the tool locally, to generate a local `index.html`, see https://github.com/protolambda/forkdiff/  (make sure to pull latest upstream commits in local repository checkout)

tynes commented:
Thoughts on squashing this into a single commit before merge? Or we just do squash merges over here?

tynes commented on core/txpool/legacypool/legacypool.go:
```diff
@@ -1739,9 +1742,26 @@ func (pool *LegacyPool) demoteUnexecutables() {
 		}
-		pendingGauge.Dec(int64(len(olds) + len(drops) + len(invalids)))
+
+		conditionalDrops, err := list.FilterTransactionConditionals(pool.currentState)
```
This runs on every new block? That could be pretty expensive in the worst case especially with 2s blocktime. Not sure what the best semantics are here to be honest, it could be to run this every 10 blocks or something like that? Curious what you think @protolambda

protolambda replied:
Every 10 blocks would hardcode the block time.

I think the ideal semantics, if it were an external thing, to look at the block witness diff, and invalidate txs quickly that way, without having to check every candidate tx storage slot. But inverting the diff like that, and accessing witness data, is going to be expensive and large diff. 

Also, doing everything only every N blocks results in lag spikes. Ideally we spread out the work.

Maybe just randomize the filter check, to sample the pool contents, to reduce workload without making it too manipulable or spiky?

hamdiallam replied:
So I ended up just removing this filter. My rationale here is that the miner will ultimately already check this when building the block and drop failed conditionals. I actually don't see the value add in trying to filter this again prior. Results in more checks for conditionals that pass (again at the miner) and doesn't reduce the number of checks for conditionals that fail (miner would filter anyways)

what do yall think about this?

tynes replied:
Are we sure that when the miner drops the txs, they leave the mempool? Does the miner operate with copies of the txs?

hamdiallam replied:
Ahh good catch! the miner operates on a copy. Didn't realize `demoteUnexecutables` was also a method of removing from the invalid txs from the mempool even when the same node has built the block. My initial understanding was that if it received a new block via p2p, this was a means of evicting bad txs.

will place it back in!

hamdiallam replied:
The main scenario to handle here is when the miner drops a tx due to a failed conditional. If the conditional passes, the transaction will depart the mempool like any other transaction in `demoteUnexecutable`.

Instead of rechecking the conditional to detect transactions that enter a failed state from mempool -> miner, we could instead set a boolean reference on the conditional marking it as rejected. Then the mempool only needs to load this reference and not recheck the whole thing.

Will be good to track the metrics to see what % of txs race to be rejected by the miner. I'd anticipate that most invalidated conditional would fail at the submission stage

hamdiallam replied:
@tynes see the latest changes that does the above. `LazyTransaction` fetches transactions from the pool by reference so we're good to directly set the `Rejected` field on the conditional

hamdiallam replied:
Added a unit test for this as well in `worker_test`. Cheap to do every block for txs that have a conditional attached to them

tynes commented on core/types/transaction_conditional.go:
```diff
@@ -0,0 +1,256 @@
+type KnownAccounts map[common.Address]KnownAccount
+
+// EncodeRLP will encode the known account into rlp bytes
+func (ka KnownAccounts) EncodeRLP(w io.Writer) error {
```
Why do we need RLP encoding of the known accounts?

hamdiallam replied:
RLP encoding is here so that the metadata is gossiped along with the tx

hamdiallam replied:
might end up removing this depending on discussion

tynes commented on eth/protocols/eth/protocol.go:
```diff
@@ -111,9 +111,86 @@ func (p *NewBlockHashesPacket) Unpack() ([]common.Hash, []uint64) {
+	return rlp.Encode(w, txs)
+}
+
+func (p *TransactionsPacket) DecodeRLP(s *rlp.Stream) error {
```
I don't think we should be sending this via P2P, it would need to be part of the protocol spec and implemented in every EL client, without knowing its something that has real usage long term I don't think its worth enshrining

hamdiallam replied:
The reason for this is so that the tx is gossiped amongst peers in the replica set that manages a tx pool. Otherwise we'll need an connection directly to the active sequencer which opens a DoS vector.

Do you think we should send directly to the active sequencer avoiding the need for gossip support?

hamdiallam commented:
thanks for the feedback! Addressed most of them and generating the forkdiff

In regards to the changes related to the transactionEnvelope that would require the specs PR there are two options
1. Make it a formal spec such that all EL clients need to do this.
2. In the proxy use the op-conductor api to determine the active sequencer and route the transaction directly. We would filter these transactions from being gossiped to avoid reverted txs in the event there's an active sequencer change and the conditional is erased and included when gossiped to the new active replica (causing a potential revert)

cc @tynes the change was needed since we rely on tx gossip between the sequencer replicas

Curious which option you guys think makes the most sense. I think (2) so that things are completely out of protocol and we an enshrine later. This does open DoS concerns with the active sequencer but we have the rate limits in place at both the proxy layer and in op-geth with @protolambda ’s suggestion

hamdiallam committed:
eth_sendRawTransactionConditional

hamdiallam commented:
Okay so ended up going with (2). Conditional transactions are bound to the block builder they were submitted to.

Removed all the rlp encoding logic and reverted the change to `TransactionPacket`. Added an check to the broadcast logic that filters out transactions with conditionals from being broadcasted.

For an architecture where different replicas can be the active sequencer, we can have the proxy simply broadcast the conditional transaction to all the replicas instead of an EL change to gossip these txs

hamdiallam commented:
soft bump @protolambda @tynes

tynes commented on core/state/statedb.go:
```diff
@@ -368,6 +368,38 @@ func (s *StateDB) HasSelfDestructed(addr common.Address) bool {
+// ensure no state changes occur while this check is executed.
+func (s *StateDB) CheckTransactionConditional(cond *types.TransactionConditional) error {
+	cost := cond.Cost()
+	if cost > types.TransactionConditionalMaxCost {
```
Does the cli flag modify this global `types.TransactionConditionalMaxCost` ?

hamdiallam replied:
it doesn't. could make it configurable but decided not to since the spec described 1k and it's other implementations hardcoded the same.

do you think it's worth making this configurable?

tynes replied:
I think its fine to not make configurable

tynes commented:
I am more open to the idea of sending the conditional requests via p2p even tho there is no way to make sure they are not malleated

tynes commented on cmd/utils/flags.go:
```diff
@@ -908,6 +908,18 @@ var (
+		Category: flags.RollupCategory,
+		Value:    false,
+	}
+	RollupSequencerTxConditionalRateLimitFlag = &cli.Uint64Flag{
```
Is this flag used? see https://github.com/ethereum-optimism/op-geth/pull/330/files#r1690167464

hamdiallam replied:
yea this flag is consumed in the miner config. It sets `RollupTransactionConditionalBurstRate` in the miner config which is forwarded to the miner rater limiter.

By having a rate limiter in the miner, it also allows a chain operator to safely expose this endpoint publicly without the proxy if they choose to

tynes replied:
I don't buy that this rate limiter would actually make this safe to expose to the public internet, if a real ddos attack makes it to userland then you are ngmi

hamdiallam replied:
yea I agree it's not sufficient but provides some level of protection to bursts of conditional txs

tynes commented on internal/sequencerapi/api.go:
```diff
@@ -0,0 +1,94 @@
+		}
+	}
+
+	state, header, err := s.b.StateAndHeaderByNumber(context.Background(), rpc.LatestBlockNumber)
```
If we use `rpc.LatestBlockNumber` here then it will create an incentive for MEV bots to only use this endpoint. Other implementations use state from a few blocks in the past, i recommend we do the same thing. We could be simple here and hardcode a random number like `tip - 3` or allow for it to be configurable

hamdiallam replied:
So we already check the conditional against `tip` and `tip-1` (line 70 of this file) to disincentivize MEV on this endpoint. Open to making this depth configurable as well

tynes replied:
Why do we check twice instead of just checking once on the older state?

hamdiallam replied:
ahh good catch, yea the state only needs to be checked on the parent. Adjusting that

tynes commented:
Curious what you think @protolambda about the P2P part of this that was previously implemented

tynes commented:
> Curious what you think @protolambda about the P2P part of this that was previously implemented

Looks like we don't need it due to changes to `proxyd` where it can forward txs to multiple endpoints

hamdiallam committed:
lint

hamdiallam commented:
soft bump @tynes @protolambda

tynes commented on core/txpool/legacypool/legacypool.go:
```diff
@@ -1723,6 +1723,7 @@ func (pool *LegacyPool) demoteUnexecutables() {
 		}
 		balance := pool.currentState.GetBalance(addr)
 		balance = pool.reduceBalanceByL1Cost(list, balance)
+
```
nit: unnecessary diff

tynes commented on core/types/transaction_conditional.go:
```diff
@@ -0,0 +1,137 @@
+			cost += len(slots)
+		}
+	}
+	if cond.BlockNumberMin != nil || cond.BlockNumberMax != nil {
```
This doesn't seem priced properly in the sense where 1 or 2 operations costs 1 unit of cost. It can be either `> min`, `< max` (both 1 operation) or `> min && < max` which is 2 operations

tynes replied:
I don't think this is a blocker or needs to be modified, just thinking out loud

hamdiallam replied:
When i was first looking over this cost calculated. It looked like it was correlated with the number of storage lookups and not comparison operations. But timestamp & block are accounted for separately even though the lookup for the block is just 1 cost so not entirely accurate either.

axelKingsley replied:
Since this is not a request for change and is a little old, i'm marking this thread resolved to clean things up.

tynes commented on internal/sequencerapi/api.go:
```diff
@@ -0,0 +1,97 @@
+
+	cost := cond.Cost()
+	sendRawTxConditionalCostMeter.Mark(int64(cost))
+	if cost > types.TransactionConditionalMaxCost {
```
constants like `TransactionConditionalMaxCost` usually live in the `params` package rather than the `types` package

tynes commented on miner/worker.go:
```diff
@@ -288,6 +295,29 @@ func (miner *Miner) commitTransaction(env *environment, tx *types.Transaction) e
+			return fmt.Errorf("failed header check: %s: %w", err, errTxConditionalInvalid)
+		}
+		if err := env.state.CheckTransactionConditional(conditional); err != nil {
+			conditional.Rejected.Store(true)
```
This atomic is a good idea, although I am not sure if the txs are copied when being pulled into the miner from the mempool, meaning that this may not make it back to the mempool, do you know if this is the case?

tynes replied:
Ok i see your comment here: https://github.com/ethereum-optimism/op-geth/pull/330#discussion_r1706194211

Disregard this

tynes commented on miner/worker.go:
```diff
@@ -423,6 +453,18 @@ func (miner *Miner) commitTransactions(env *environment, plainTxs, blobTxs *tran
+		case errors.Is(err, errTxConditionalInvalid):
+			// err contains contextual info on the failed conditional
+			txConditionalRejectedCounter.Inc(1)
+			log.Debug("Skipping account, transaction with failed conditional", "sender", from, "hash", ltx.Hash, "err", err)
```
could set atomic rejected to true here rather than having it twice upstream of here

tynes replied:
but also have the context of the conditional upstream so there are tradeoffs, not super opinionated

hamdiallam replied:
originally had it above to not set `Rejected` on rate limited errors. But that's handled separately in this switch. I like the suggestion

tynes commented on internal/sequencerapi/errors.go:
```diff
@@ -0,0 +1,20 @@
+package sequencerapi
+
```
We need this to get access to the error code? Its easier to introduce something rather than modify the `rpc.Error`?

hamdiallam replied:
ah yea good point, making `rpc.JsonError` public would be sufficient and better. Also very useful when using the rpc library externally in other services as well

tynes commented:
Generally looks good to me, just a few open nits/questions

hamdiallam committed:
feedback

hamdiallam committed:
nits

tynes commented on core/types/transaction_conditional.go:
```diff
@@ -0,0 +1,130 @@
+	TimestampMax   *uint64  `json:"timestampMax,omitempty"`
+
+	// Internal fields used for metrics and inclusion tracking
+	SubmissionTime time.Time    `json:"-"`
```
This is decorated by the proxy? There is already a time property on the transaction:
- https://github.com/ethereum-optimism/op-geth/blob/60038121c7571a59875ff9ed7679c48c9f73405d/core/types/transaction.go#L58
- https://github.com/ethereum-optimism/op-geth/blob/60038121c7571a59875ff9ed7679c48c9f73405d/core/types/transaction.go#L561

hamdiallam replied:
ah no it's not, didn't realize this was available! I'll edit to use this instead.

hamdiallam replied:
Looking at the docstring and usage. It looks like this is only used in tests

Going to update the docstring as well and set the time in the API handler

tynes replied:
It is used in the block building to break ties on same fee - https://github.com/ethereum/go-ethereum/blob/d71831255da7ecb3817c9e14c9142fe9d4441d3b/miner/ordering.go#L67

hamdiallam replied:
ahh okay I see, thanks

tynes commented on core/types/transaction_conditional.go:
```diff
@@ -0,0 +1,130 @@
+
+	// Internal fields used for metrics and inclusion tracking
+	SubmissionTime time.Time    `json:"-"`
+	Rejected       *atomic.Bool `json:"-"`
```
We are interested in moving this `Rejected` field off of the `TransactionConditional` struct and to the `Transaction` struct as part of interop, so that the same field can be used to validate executing messages

cc @protolambda

protolambda replied:
+1

hamdiallam replied:
Nice, would y'all want to update this later to move into `Transaction` or thinking of updating this PR to jumpstart that abstraction?

protolambda replied:
Ideally in this PR

hamdiallam replied:
okay sounds good! I think it actually makes the code a bit cleaner when it's in `Transaction`. I had to make `Rejected` a pointer to an atomic since I wanted the conditional to be passed by value. Felt a bit awkward to init the point at the api layer when forwarding the tx

protolambda commented on rpc/json.go:
```diff
@@ -136,24 +136,24 @@ func errorMessage(err error) *jsonrpcMessage {
 }
 
-type jsonError struct {
+type JsonError struct {
```
Not entirely sure, but I think we could add a Go type alias (not a new type), to expose it, without changing the code of every receiver method here. That reduces the diff and improves maintainability.

Like so:
```go
type JsonError = jsonError
```
The `=` makes it the same underlying type, not a new type.

hamdiallam replied:
neat! Didn't realize you can do this. Let me try

hamdiallam committed:
JsonError type alias

hamdiallam committed:
use tx.Time() instead of manual tracking

protolambda commented on cmd/utils/flags.go:
```diff
@@ -976,6 +976,18 @@ var (
+		Category: flags.RollupCategory,
+		Value:    false,
+	}
+	RollupSequencerTxConditionalRateLimitFlag = &cli.Uint64Flag{
```
```suggestion
	RollupSequencerTxConditionalRateLimitFlag = &cli.IntFlag{
```
So that it matches the type of the config

protolambda commented on cmd/utils/flags.go:
```diff
@@ -1717,6 +1729,9 @@ func setMiner(ctx *cli.Context, cfg *miner.Config) {
 		cfg.RollupComputePendingBlock = ctx.Bool(RollupComputePendingBlock.Name)
 	}
+	if ctx.IsSet(RollupSequencerTxConditionalRateLimitFlag.Name) {
+		cfg.RollupTransactionConditionalBurstRate = int(ctx.Uint64(RollupSequencerTxConditionalRateLimitFlag.Name))
```
```suggestion
		cfg.RollupTransactionConditionalBurstRate = ctx.Int(RollupSequencerTxConditionalRateLimitFlag.Name)
```
See CLI flag comment; to match the type of the config

protolambda commented on miner/miner.go:
```diff
@@ -84,18 +87,22 @@ type Miner struct {
+		txpool:             eth.TxPool(),
+		chain:              eth.BlockChain(),
+		pending:            &pending{},
+		conditionalLimiter: rate.NewLimiter(params.TransactionConditionalMaxCost, config.RollupTransactionConditionalBurstRate),
```
If you add a comment line above the  `conditionalLimiter` field then the diff becomes smaller, and you can high-light that this field is an op-stack modification, such that we now what to do with it on future Geth merges.

protolambda commented on fork.yaml:
```diff
@@ -268,6 +268,25 @@ def:
+            - "eth/protocols/eth/broadcast.go"
+            - "internal/sequencerapi/api.go"
+            - "internal/sequencerapi/errors.go"
+            - "miner/worker.go"
+            - "miner/miner.go"
```
Forkdiff doesn't support duplicates yet (`miner.go` and `worker.go` were modified in other places). Adding a note in those places of the forkdiff yaml, that there's a change to support `eth_sendRawTransactionConditional`, could resolve this.

hamdiallam replied:
sounds good, I realize I also need to regenerate this file too with some of the recent changes

protolambda commented on eth/backend.go:
```diff
@@ -374,6 +375,16 @@ func (s *Ethereum) APIs() []rpc.API {
 
+	// Append any Sequencer APIs as enabled
+	if s.config.RollupSequencerEnableTxConditional {
+		// an effort to detect if running in a non-sequencer mode
+		if len(s.config.RollupSequencerHTTP) > 0 {
+			log.Warn("SendRawTransactionConditional enabled when RollupSequencerHTTP is defined. Skipping...")
```
The comment is unclear. If the `RollupSequencerHTTP` is set, then it assumes a sentry setup, and doesn't enable the conditional checks? Or vice versa? I think it's a bad idea to change the behavior. A warning log is nice, but the conditional checks and API should always be enabled if the flag is set to true.

hamdiallam replied:
Good q, curious to get your thoughts

So in the current implementation, it does **not** support have support for forwarding `eth_sendRawTranscationConditional` requests to the sequencer after validation. The caller needs to send the request directly to the sequencer. So this code block here was an effort to not add the api to the node if it wasn't the sequencer

Looking at this again though since we already have `RollupSequencerHTTP` maybe it's worth adding support for forwarding these requests as it's a small lift in the code. Note it's not the same as forwarding txs since we want to explicitly call `sendRawTransactionConditional` and not `sendRawTransaction` which would strip the conditional

hamdiallam replied:
updated the implementation such that conditional txs can be forwarded to the sequencer if the `seqRPCService` is set.

This is the same client used by `SendTx` so we won't end up in a situation where the conditional is lost and wrongly propagated to the sequencer. & if it was a different client from `SendTx`, the clients are only created when `RollupSequencerHTTP` is set adding another layer of protection here

hamdiallam replied:
LMK how the updated code looks

protolambda commented on core/state/statedb.go:
```diff
@@ -419,6 +419,38 @@ func (s *StateDB) HasSelfDestructed(addr common.Address) bool {
+// ensure no state changes occur while this check is executed.
+func (s *StateDB) CheckTransactionConditional(cond *types.TransactionConditional) error {
+	cost := cond.Cost()
+	if cost > params.TransactionConditionalMaxCost {
```
Is the `max` supposed to be inclusive or exclusive?

hamdiallam replied:
included a comment about it being inclusive

protolambda commented on params/conditional_tx_params.go:
```diff
@@ -0,0 +1,8 @@
+package params
+
+const (
+	TransactionConditionalMaxCost = 1000
```
Worth a comment if this is an inclusive or exclusive maximum.

protolambda commented on internal/sequencerapi/api.go:
```diff
@@ -0,0 +1,92 @@
+		}
+	}
+
+	// State is checked against an older block to remove the MEV incentive for this endpoint compared with sendRawTransaction
+	parentBlock := rpc.BlockNumberOrHash{BlockHash: &header.ParentHash}
+	parentState, _, err := s.b.StateAndHeaderByNumberOrHash(context.Background(), parentBlock)
```
I am not sure if this is worth it, as we do want accurate error information to apply back-pressure to bad actors who spam.

hamdiallam replied:
Not worth checking against past state at all? How else might we discourage this endpoint against MEV vs sendRawTransaction?

tynes replied:
@protolambda what do you mean exactly? Are you saying the state check should only happen in the sequencer? You are right that there are consistency issues here, its a tradeoff, I think its better to do the check here, because really only high contention environments will break, which is ok because the searchers shouldn't use this endpoint

protolambda replied:
I meant the usage of a parent block, instead of the already known latest block, here. The MEV question is interesting. Using older state might backfire more than we desire. How do other protocols handle this MEV edge? Is there a standard?

I'm fine with merging this as-is, but this parent-block choice is worth some discussion async.

protolambda commented on core/types/transaction.go:
```diff
@@ -64,6 +64,9 @@ type Transaction struct {
 
 	// cache of details to compute the data availability fee
 	rollupCostData atomic.Value
+
+	// optional preconditions for inclusion
+	conditional atomic.Value
```
Need to double-check the Go version, but I believe the std-lib has Go generics for `atomic`, to not have to type-check on every Load.

protolambda commented on core/types/transaction_conditional.go:
```diff
@@ -0,0 +1,130 @@
+	return nil
+}
+
+// Cost computes the aggregate cost of the preconditions; total number of storage lookups required
+func (cond *TransactionConditional) Cost() int {
+	cost := 0
+	for _, account := range cond.KnownAccounts {
+		if _, isRoot := account.Root(); isRoot {
+			cost += 1
+		}
+		if slots, isSlots := account.Slots(); isSlots {
+			cost += len(slots)
+		}
+	}
```
Is this cost function standardized somewhere? I think it makes sense to always add `1` for an account, even if the account is empty, or if there are slots but no root. Since we end up doing a little bit of work per known account entry. E.g. we wouldn't want a super-long list of accounts that we all count as 0 cost, because the accounts-list still fills up memory.

tynes replied:
I do not believe it is standardized, it was one thing that I thought was lacking in the original spec

hamdiallam replied:
adding +1 as a default per account sounds good to me. Can adjust later on

protolambda reviewed (commented):
Reviewed everything except the test-cases, will revisit those tomorrow.

hamdiallam committed:
some feedback

hamdiallam committed:
lift rejected into transaction

hamdiallam committed:
support for forwarding sendRawTransactionConditional to seqRPC

anacrolix commented on miner/worker.go:
```diff
@@ -287,6 +294,27 @@ func (miner *Miner) commitTransaction(env *environment, tx *types.Transaction) e
+		}
+
+		// ditch the reservation as we've checked that we're allowed `cost` units by the limiter
+		miner.conditionalLimiter.ReserveN(now, cost)
```
I'm not sure this looks right. You are checking if there is enough tokens to proceed immediately, but erroring right away. Then you reserve them, presumably there's a race here.

I think you want to take the ReserveN, check if it's available immediately, and if it isn't, cancel the reservation so someone else can have it.

hamdiallam replied:
the miner runs in a single threaded loop so this should eliminate the race condition right?

But ah, didn't see that a reservation could be canceled, that actually feels cleaner! When implementing it looked like the holder would have to block once taking the rez. Thanks for pointing that out

anacrolix replied:
Not a data race. Consider if multiple requests passed the Allow check, and then all dropped the Reserve. What would happen is a huge burst that would disallow any new requests for quite sometime.

If it behaved as you described, after they all passed the allow, some of them would actually block, which also not what I think you intended.

I still don't think it's right, the rate package API is pretty confusing:

I think you want to take the reservation. Checking it's OK only tells you if it's possible to satisfy the token count you requested before the end of the universe (1<<63 nano seconds after 1970). You still need to check what the delay is, so check for `reservation.Delay() > 0`. If it's true, cancel your reservation (again assuming you don't want to block...), `reservation.Cancel()`. Anyone that gets `reservation.Delay==0` is ready to proceed immediately.

anacrolix replied:
I didn't want to push on top of your branch unless you want:

https://github.com/ethereum-optimism/op-geth/compare/sendRawTransactionConditional...anacrolix/sendRawTransactionConditional

I also spotted that the rate Limiter parameters were passed in the wrong order.

hamdiallam replied:
ahh yea good catch, thanks for the explainer

hamdiallam replied:
> I also spotted that the rate Limiter parameters were passed in the wrong order.

Hmm, I intentionally had it that order but maybe misunderstanding the API. My thinking was that the limiter is refilled at a fixed rate 1000 units per second. And we allow the bucket size to be configurable. So for example with a  value of 5000, if 1000 units was consumed with each conditional, this configuration would only allow 5 conditional txs in a given second.

With the arguments reversed, only 1000 units could ever be consumed in a given second. This is understanding correct?

anacrolix replied:
Ah, I'm bummed you didn't just take the commit.

You currently have it so that 1000 units are available per second, with a "debt" (burst) of `RollupTransactionConditionalBurstRate`. Incidentally you actually have `RollupTransactionConditionalBurstRate` set to 1000 too, FWIW.

hamdiallam replied:
AHH wait okay i understand now, wow this is confusing.

The burst is just in a given call. So your original suggestion of swapping the two makes sense!! Should've just taken the commit 😅

anacrolix commented on internal/sequencerapi/api.go:
```diff
@@ -0,0 +1,96 @@
+	}
+}
+
+func (s *sendRawTxCond) SendRawTransactionConditional(ctx context.Context, txBytes hexutil.Bytes, cond types.TransactionConditional) (common.Hash, error) {
```
A reference to the contents of cond is taken in an atomic.Pointer. It's not ideal, but it just happens to work because the receiver gets wrapped in the RPC API, which seems to generate owned values and hand them over wholesale to the method here. I can see that TransactionConditional contains pointers for optional values so that must be why it's done like this.

Thought I should flag it in case it wasn't intentional, but I think it's fine.

hamdiallam replied:
yeah exactly, not ideal but a pattern in go that the transaction is passed around pretty much everywhere by reference and the optional values are populated prior to creating that transaction, then handed off via the atomic pointer on the transaction itself.

Open to changing this though if you feel there's a better method

anacrolix replied:
I think it's out of scope for this PR, I think it's going to be the default through the RPC code.

hamdiallam replied:
okay sounds good, going to resolve this thread

anacrolix reviewed (approved):
Seems good to me. I dug into the RPC API to check some of the assumptions.

I don't think the rate limiting is quite right, but I don't know if the caller should be handling their allowed rate, or if it should be blocking in the RPC (I guess the former).

There's a lot of use of atomics which can mask concurrency bugs but that seems to be widespread in op-geth.

hamdiallam committed:
feedback on limiter reservation. log skipped conditional txs at warn to start

hamdiallam committed:
update forkdiff

hamdiallam committed:
limiter reservation delay needs to be checked

tynes commented on core/types/transaction_conditional.go:
```diff
@@ -0,0 +1,126 @@
+// then the user prefers their transaction to only be included if the
+// particular storage slot values from state match.
+type KnownAccount struct {
+	StorageRoot  *common.Hash `rlp:"nil"`
```
Does this need `rlp:"nil"` still?

hamdiallam replied:
ah nope, good find

tynes commented on internal/sequencerapi/api.go:
```diff
@@ -0,0 +1,100 @@
+		}
+	}
+
+	header, err := s.b.HeaderByNumber(context.Background(), rpc.LatestBlockNumber)
```
Shouldn't we be consistent with the height used for the header check and the state check? The client side would likely assume that they are consistent. The only way that I could imagine this making sense is if there is no overlap between usage of the header check and the state check. This detail needs product input.

I would recommend we make it consistent for now to not slow down this PR and then we can follow up with talking to some bundlers to learn more.

We could also look at other implementations and try to be consistent with them.

hamdiallam replied:
ah yes ur right.

I thought it'd be an optimization to only check against the parent since but state could be invalidated between parent/head

tynes commented on miner/worker.go:
```diff
@@ -287,6 +294,31 @@ func (miner *Miner) commitTransaction(env *environment, tx *types.Transaction) e
 		return miner.commitBlobTransaction(env, tx)
 	}
+
+	// If a conditional is set, check prior to applying
```
Have you considered putting the rate limiter in the RPC layer? Just curious if you have thought about it at all. It seems like it would be better UX for bundlers to learn that their transactions is not gonna make it earlier rather than it being silently dropped in the miner

tynes replied:
It also saves some resources as the transaction would be rejected directly higher up in the stack. Now that I think about it more, i feel strongly that its a better idea to put this in the RPC layer

anacrolix replied:
It makes sense if the intention is not to arbitrarily abort transactions this far down, but does it expand the scope of the PR? Can it be done in a later PR rather than block this one?

hamdiallam replied:
The check is a last ditch effort for transactions that race and invalidate each other in the same block. We have metrics here to detect how often this happens.

But this check is absolutely done in the RPC layer, `sequencerapi`. And also there's support for sentry nodes to do this check as well since these checks also occur there prior to forwarding to the sequencer url.

In order to completely do it in the RPC layer, there would have to be a way to track pending conditional txs that would invalidate each other intrablock which would be pretty complex to implement

hamdiallam replied:
So the rate limiter is actually also present in the RPC layer. See [here](https://github.com/ethereum-optimism/infra/pull/42/files#diff-5adc79c515f1776136f9b28176617ee4e852caddf0f637a5c733c768c730880bR11)

So by running the tx proxy, the limiter in the block builder should effectively be a no-op. It was added based on a suggestion by @protolambda for an additional layer of safety. It also allows other chains that just want to enable this without the proxy to sort-of do so with some level of safety within op-geth as well

anacrolix commented on miner/worker.go:
```diff
@@ -287,6 +294,31 @@ func (miner *Miner) commitTransaction(env *environment, tx *types.Transaction) e
+		now, cost := time.Now(), conditional.Cost()
+		res := miner.conditionalLimiter.ReserveN(now, cost)
+		if !res.OK() {
+			res.Cancel()
```
Pedantic but this isn't necessary.

https://pkg.go.dev/golang.org/x/time/rate#Reservation.OK

https://github.com/ethereum-optimism/op-geth/compare/sendRawTransactionConditional...anacrolix/sendRawTransactionConditional#diff-689874b30f6f905ce6c47cdefeddcf052b467a7936a981f4b6cf38939e10d924R302-R311

https://cs.opensource.google/go/x/time/+/refs/tags/v0.6.0:rate/rate.go;l=169-171

hamdiallam replied:
ah nice

hamdiallam committed:
feedback

hamdiallam committed:
feedback on limit. lol sorry for the originally taking the commit!

axelKingsley commented on core/state/statedb_test.go:
```diff
@@ -1373,3 +1373,221 @@ func TestStorageDirtiness(t *testing.T) {
+	}
+
+	tests := []struct {
+		name       string
+		preActions []preAction
+		cond       types.TransactionConditional
+		valid      bool
+	}{
+		{
+			// Clean Prestate, no defined cond
+			name:       "clean prestate",
+			preActions: []preAction{},
+			cond:       types.TransactionConditional{},
+			valid:      true,
+		},
```
It should be possible to omit the variable names for an implied structure like this, which might make the test structure smaller. But maybe not easier to understand.

hamdiallam replied:
yea my editor yells at me that these variables don't have to be named but it makes the tests cleaner imo

axelKingsley commented on core/txpool/legacypool/legacypool.go:
```diff
@@ -1733,16 +1733,26 @@ func (pool *LegacyPool) demoteUnexecutables() {
+		})
+		for _, tx := range rejectedDrops {
+			hash := tx.Hash()
+			log.Trace("Removed rejected transaction by the miner", "hash", hash)
+			pool.all.Remove(hash)
```
Technically the trace is not accurate until after `pool.all.Remove`. By no means blocking, as this appears not to have a chance of error.

axelKingsley commented on miner/worker.go:
```diff
@@ -287,6 +294,30 @@ func (miner *Miner) commitTransaction(env *environment, tx *types.Transaction) e
+
+		txConditionalMinedTimer.UpdateSince(tx.Time())
+
+		// check the conditional
+		if err := env.header.CheckTransactionConditional(conditional); err != nil {
+			return fmt.Errorf("failed header check: %s: %w", err, errTxConditionalInvalid)
+		}
+		if err := env.state.CheckTransactionConditional(conditional); err != nil {
+			return fmt.Errorf("failed state check: %s: %w", err, errTxConditionalInvalid)
+		}
```
Note to self (and others): This is the core application of the conditional logic when applying transactions.

axelKingsley commented on core/state/statedb.go:
```diff
@@ -425,6 +425,40 @@ func (s *StateDB) HasSelfDestructed(addr common.Address) bool {
+		return fmt.Errorf("conditional cost, %d, exceeded max: %d", cost, params.TransactionConditionalMaxCost)
+	}
+
+	for addr, acct := range cond.KnownAccounts {
+		if root, isRoot := acct.Root(); isRoot {
+			storageRoot := s.GetStorageRoot(addr)
+			if storageRoot == (common.Hash{}) { // if the root is not found, replace with the empty root hash
+				storageRoot = types.EmptyRootHash
+			}
+			if root != storageRoot {
+				return fmt.Errorf("failed account storage root constraint. Got %s, Expected %s", storageRoot, root)
+			}
+		}
+		if slots, isSlots := acct.Slots(); isSlots {
+			for key, state := range slots {
+				accState := s.GetState(addr, key)
+				if state != accState {
+					return fmt.Errorf("failed account storage slot key %s constraint. Got %s, Expected %s", key, accState, state)
+				}
+			}
+		}
+	}
+	return nil
```
Note to self (and others): this is the core section of code which actually does Conditional computation.

axelKingsley reviewed (approved):
I looked over this and I think we should prioritize shipping it soon. There are some open questions, and some resolved ones that I closed.

@protolambda - please confirm this is not blocking: https://github.com/ethereum-optimism/op-geth/pull/330/files#r1751043303

@tynes and @anacrolix - I see we've got an approval from Matt, and then some further comments, but none of them appeared blocking. Is there anything critical we need done before merging this?

Looks like the API interface for this comes into `txproxy` here: https://github.com/ethereum-optimism/infra/pull/42

And there's a monorepo E2E test for this feature: https://github.com/ethereum-optimism/optimism/pull/11671/files

Given the E2E test passes for @hamdiallam , and it's ready for its own review, I feel comfortable with the changes.

If there's still a lot of activity here because we believe this feature isn't ready to land, I want to make sure we know which conversations are critical. Since I don't see any, I want to ask for some 👍 emojis so we can merge this.

tynes commented on internal/sequencerapi/api.go:
```diff
@@ -0,0 +1,106 @@
+
+	state, header, err := s.b.StateAndHeaderByNumber(context.Background(), rpc.LatestBlockNumber)
+	if err != nil {
+		return common.Hash{}, err
```
I am still confused as to why we check the state twice. Shouldn't we do the quickest lookup to the chain tip then grab an ancestor and check both header and state against that ancestor? There is also the edge case of a short chain

hamdiallam replied:
Against HEAD we check header/state. Header is checked here since I don't see why block inclusion conditions (number/timestamp) needs to be stale to prevent MEV. State is also checked here since the transition from ancestor -> head could have invalidated the condition for account state. We could opt to not check state here and let this fail later in the pipeline at the miner.

For the ancestor, we only check the state since that's the only thing that MEV searchers would be comparing against? So the state check should also be stale


Does this help clear up the intent here? Open to not checking against HEAD and letting it fail in the miner for the edge case that ancestor -> head would have invalidated the conditional

tynes commented on internal/sequencerapi/api.go:
```diff
@@ -0,0 +1,106 @@
+		}
+	}
+
+	state, header, err := s.b.StateAndHeaderByNumber(context.Background(), rpc.LatestBlockNumber)
```
Should we use the same `ctx` that is passed into the top level function?

hamdiallam replied:
ah yes let me do that

tynes commented on internal/sequencerapi/api.go:
```diff
@@ -0,0 +1,106 @@
+
+	// forward if seqRPC is set, otherwise submit the tx
+	if s.seqRPC != nil {
+		var hash common.Hash
```
One thing that i am noticing is that we don't do the fee check in this branch

https://github.com/ethereum-optimism/op-geth/blob/60038121c7571a59875ff9ed7679c48c9f73405d/internal/ethapi/api.go#L1997

hamdiallam replied:
oh good call, I can add these preconditions here as well. Not ideal to dup that code but an alternative approach would be a bigger diff

hamdiallam committed:
feedback

hamdiallam committed:
small log on enabling sendRawTransactionConditional
